File: backend/src/readaloud/services/reading_cues.py
"""Sentence-level timing cues for synced reading highlights.

Cue times live in the stitched audio's own timeline, so they are
unaffected by playback-rate changes the frontend applies at playback time.
"""

import re

from readaloud.models.schemas import Cue
from readaloud.services.text_chunker import split_sentences


def compute_cues(chunk_texts: list[str], chunk_durations: list[float]) -> list[Cue]:
    """Build sentence-level playback cues for a sequence of TTS chunks.

    Each chunk's real audio duration is split across its own sentences,
    proportional to sentence character length. Sentences never cross chunk
    boundaries, so a chunk produced by splitting an oversized sentence
    across multiple TTS calls (see `text_chunker._split_long_sentence`)
    yields one cue per chunk piece rather than trying to recombine them.

    Args:
        chunk_texts: Chunk text, in playback order -- the same list passed
            to the TTS server for synthesis.
        chunk_durations: Each chunk's real audio duration in seconds, same
            order and length as `chunk_texts`.

    Returns:
        Cues in playback order, with `start`/`end` in the stitched audio's
        timeline (seconds).

    Raises:
        ValueError: If `chunk_texts` and `chunk_durations` differ in length.
    """
    if len(chunk_texts) != len(chunk_durations):
        raise ValueError("chunk_texts and chunk_durations must be the same length")

    cues: list[Cue] = []
    offset = 0.0
    for text, duration in zip(chunk_texts, chunk_durations, strict=True):
        cues.extend(_cues_for_chunk(text, duration, offset))
        offset += duration
    return cues


def _cues_for_chunk(text: str, duration: float, offset: float) -> list[Cue]:
    paragraphs = [p for p in re.split(r"\n\n+", text) if p.strip()]
    sentences_by_paragraph = [[s for s in split_sentences(p) if s.strip()] for p in paragraphs]
    all_sentences = [s for para in sentences_by_paragraph for s in para]
    total_chars = sum(len(s) for s in all_sentences)
    if not all_sentences or total_chars == 0:
        return []

    cues: list[Cue] = []
    start = offset
    for para_index, para_sentences in enumerate(sentences_by_paragraph):
        for sent_index, sentence in enumerate(para_sentences):
            end = start + duration * (len(sentence) / total_chars)
            is_last_sentence_in_para = sent_index == len(para_sentences) - 1
            is_last_paragraph = para_index == len(sentences_by_paragraph) - 1
            suffix = "\n\n" if is_last_sentence_in_para and not is_last_paragraph else ""
            cues.extend(_cues_for_sentence(sentence, start, end, suffix))
            start = end
    return cues


def _cues_for_sentence(sentence: str, start: float, end: float, suffix: str) -> list[Cue]:
    """Split one sentence's allotted time span across its words.

    Words share the span proportional to their character length, the same
    proportional approach `_cues_for_chunk` uses for sentences -- there is no
    real per-word timing from the TTS server to split on.
    """
    words = sentence.split()
    total_chars = sum(len(word) for word in words)
    if not words or total_chars == 0:
        return []

    duration = end - start
    cues: list[Cue] = []
    cursor = start
    for index, word in enumerate(words):
        word_end = cursor + duration * (len(word) / total_chars)
        is_last_word = index == len(words) - 1
        cue_text = word + suffix if is_last_word else word
        cues.append(Cue(text=cue_text, start=cursor, end=word_end))
        cursor = word_end
    return cues

File: extension/lib/reading-cues.js
import { splitSentences } from "/lib/chunker.js";

/**
 * Word-level cues for one synthesized chunk.
 *
 * JS port of backend/src/readaloud/services/reading_cues.py, restricted to a
 * single chunk: times start at 0, since the extension plays each chunk as its
 * own audio element rather than one stitched file.
 */

/** Matches Python's `[^\w\s]+` against a whole token. */
const PUNCTUATION_ONLY = /^[^\p{L}\p{N}_\s]+$/u;
const MIN_TIMESTAMP_COVERAGE = 0.8;

/**
 * Build one chunk's cues, timed from 0.
 *
 * Uses the TTS server's real per-word timestamps when they line up with the
 * chunk's own words, and otherwise splits `duration` across those words
 * proportional to character length. Cue text always comes from `text`, never
 * from the timestamps -- the server reports what it spoke, which its text
 * normalizer may have rewritten.
 */
export function cuesForChunk(text, duration, timestamps) {
  if (timestamps && timestamps.length) {
    const cues = cuesFromTimestamps(text, timestamps, duration);
    if (cues) return cues;
  }
  return heuristicCues(text, duration);
}

function paragraphsOf(text) {
  return text.split(/\n\n+/).filter((paragraph) => paragraph.trim());
}

function wordsOf(text) {
  return text.split(/\s+/).filter(Boolean);
}

function heuristicCues(text, duration) {
  const sentencesByParagraph = paragraphsOf(text).map((paragraph) =>
    splitSentences(paragraph).filter((sentence) => sentence.trim()),
  );
  const allSentences = sentencesByParagraph.flat();
  const totalChars = allSentences.reduce((sum, sentence) => sum + sentence.length, 0);
  if (!allSentences.length || totalChars === 0) return [];

  const cues = [];
  let start = 0;
  sentencesByParagraph.forEach((paragraphSentences, paragraphIndex) => {
    paragraphSentences.forEach((sentence, sentenceIndex) => {
      const end = start + duration * (sentence.length / totalChars);
      const lastInParagraph = sentenceIndex === paragraphSentences.length - 1;
      const lastParagraph = paragraphIndex === sentencesByParagraph.length - 1;
      const suffix = lastInParagraph && !lastParagraph ? "\n\n" : "";
      cues.push(...sentenceCues(sentence, start, end, suffix));
      start = end;
    });
  });
  return cues;
}

function sentenceCues(sentence, start, end, suffix) {
  const words = wordsOf(sentence);
  const totalChars = words.reduce((sum, word) => sum + word.length, 0);
  if (!words.length || totalChars === 0) return [];

  const duration = end - start;
  const cues = [];
  let cursor = start;
  words.forEach((word, index) => {
    const wordEnd = cursor + duration * (word.length / totalChars);
    const isLast = index === words.length - 1;
    cues.push({ text: isLast ? word + suffix : word, start: cursor, end: wordEnd });
    cursor = wordEnd;
  });
  return cues;
}

/**
 * Fold punctuation-only tokens into the adjacent word's span.
 *
 * Kokoro tokenizes punctuation separately ("test" then "."), while the source
 * text attaches it ("test."). Punctuation merges backward, or forward when it
 * opens the chunk -- a paragraph starting on a quotation mark is common enough
 * that leaving it standalone would fail alignment for the whole chunk.
 */
function mergePunctuation(timestamps) {
  const merged = [];
  let leading = [];

  for (const stamp of timestamps) {
    const isPunctuation = PUNCTUATION_ONLY.test(stamp.word);
    if (isPunctuation && merged.length) {
      const previous = merged[merged.length - 1];
      merged[merged.length - 1] = {
        word: previous.word + stamp.word,
        start: previous.start,
        end: stamp.end,
      };
    } else if (isPunctuation) {
      leading.push(stamp);
    } else if (leading.length) {
      merged.push({
        word: leading.map((entry) => entry.word).join("") + stamp.word,
        start: leading[0].start,
        end: stamp.end,
      });
      leading = [];
    } else {
      merged.push(stamp);
    }
  }

  if (leading.length) {
    merged.push({
      word: leading.map((entry) => entry.word).join(""),
      start: leading[0].start,
      end: leading[leading.length - 1].end,
    });
  }
  return merged;
}

/** The chunk's own words in order, each paragraph's last word carrying "\n\n". */
function chunkWords(text) {
  const paragraphs = paragraphsOf(text);
  const words = [];
  paragraphs.forEach((paragraph, index) => {
    const paragraphWords = wordsOf(paragraph);
    if (index < paragraphs.length - 1 && paragraphWords.length) {
      paragraphWords[paragraphWords.length - 1] += "\n\n";
    }
    words.push(...paragraphWords);
  });
  return words;
}

/** Kokoro can report a negative start for a chunk's first word. */
function clampStart(start, duration) {
  const clamped = Math.max(0, start);
  return duration > 0 ? Math.min(clamped, duration) : clamped;
}

/**
 * Returns the chunk's cues, or null when the timestamps can't be trusted: a
 * token count that doesn't match the chunk's own words (the normalizer
 * rewrote something, so no positional mapping holds), or timestamps ending
 * well before the audio does (which would freeze the highlight).
 */
function cuesFromTimestamps(text, timestamps, duration) {
  const merged = mergePunctuation(timestamps);
  const words = chunkWords(text);
  if (!merged.length || merged.length !== words.length) return null;
  if (duration > 0 && merged[merged.length - 1].end < duration * MIN_TIMESTAMP_COVERAGE) {
    return null;
  }

  const cues = words.map((word, index) => ({
    text: word,
    start: clampStart(merged[index].start, duration),
    end: merged[index].end,
  }));
  for (let index = 0; index < cues.length - 1; index++) {
    cues[index].end = cues[index + 1].start;
  }
  if (duration > 0) cues[cues.length - 1].end = duration;
  return cues;
}
